Add money back mention in checkout page

monsieur-z opened the pull request:
### Changes proposed in this Pull Request

This PR adds a money back guarantee mention in the footer of the checkout page (for Jetpack products). It also centers the submit button on mobile.

Fixes 1196108640073826-as-1200480578254960

### Testing instructions

#### Prerequisittes
- Make sure you have a simple site and a Jetpack site available
- Download the PR and run Calypso

#### Instructions
- Visit the plans page with your simple site selected
- Add a product to your cart
- Check that you **don't** see the money back mention at the bottom of the page
- Visit the plans page with your Jetpack site selected
- Add a **yearly plan** to the cart
- Check that the submit button of the checkout page is centered on mobile
- Notice the 14 day money back mention at the bottom of the page
- Check that on mobile (viewport width < 700px) the sticky footer doesn't hide the content of the page
- Go back to the plans page and add a **monthly product**
- Check that you see both the 14 day and 7 day mentions
- Remove the yearly plan from the cart
- Check that you see the 7 day money back mention

### Screenshots

#### Money back mention

_Before_
<img width="593" alt="Screen Shot 2021-06-30 at 9 08 16 AM" src="https://user-images.githubusercontent.com/1620183/123979327-a0208000-d98e-11eb-9df6-1ca4c331eb55.png">


_After (yearly products)_
<img width="601" alt="Screen Shot 2021-06-30 at 10 01 24 AM" src="https://user-images.githubusercontent.com/1620183/123979355-a57dca80-d98e-11eb-8c99-480eeb664e96.png">

_After (monthly products)_
<img width="586" alt="Screen Shot 2021-07-07 at 10 37 33 AM" src="https://user-images.githubusercontent.com/1620183/124779718-0158ce00-df10-11eb-81cb-4c8487917e8d.png">


_After (both yearly and monthly products)_
<img width="577" alt="Screen Shot 2021-07-07 at 10 37 55 AM" src="https://user-images.githubusercontent.com/1620183/124779742-0584eb80-df10-11eb-9e2b-da9d7e57def1.png">


#### Submit button

_Before_
<img width="388" alt="Screen Shot 2021-06-30 at 10 05 19 AM" src="https://user-images.githubusercontent.com/1620183/123979380-aadb1500-d98e-11eb-8321-8ecf2e661385.png">


_After_
<img width="395" alt="Screen Shot 2021-06-30 at 10 05 06 AM" src="https://user-images.githubusercontent.com/1620183/123979398-ae6e9c00-d98e-11eb-89aa-0f237317e07a.png">

Commit:
Add money back mention in checkout page

mattgawarecki reviewed (approved):
GitHub seems to be a little unhappy for some reason, but the errors it lists seem incorrect. I tested on calypso.live and everything looked exactly as described, and the code changes look reasonable to me. 👍

sirbrillig commented on packages/composite-checkout/src/components/checkout-steps.tsx:
```diff
@@ -506,6 +526,10 @@ export function CheckoutStepArea( {
 				/>
+				<SubmitButtonFooter>
+					<img src={ badge14Src } alt="" />
+					<span>{ __( '14 day money back guarantee' ) }</span>
```
I don't think that this guarantee applies to all possible uses of the `composite-checkout` module. Although wpcom is currently the  only user of this package, it's designed with no knowledge of wpcom and is intended to be used with other services as well. I think this might do better if we put this badge in calypso (the `client` directory).

monsieur-z replied:
Thanks @sirbrillig. I'll clarify which products are impacted by this guarantee.

sirbrillig commented:
I think the compile errors you're seeing are because the `composite-checkout` npm package isn't set up to build with image imports (at least for TypeScript), so the build of that package is failing, which causes calypso (that requires that package) to fail as well.

Commit:
Move checkout footer to wpcom codebase

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx:
```diff
@@ -452,6 +454,12 @@ export default function WPCheckout( {
+				submitButtonFooter={
+					<SubmitButtonFooter>
+						<img src={ badge14Src } alt="" />
+						<span>{ translate( '14 day money back guarantee' ) }</span>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( '7 day money back guarantee' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/improve/checkout-money-back-mention)

monsieur-z replied:
I'll clarify if that's for a subset of products only.

monsieur-z commented:
> I think the compile errors you're seeing are because the `composite-checkout` npm package isn't set up to build with image imports (at least for TypeScript), so the build of that package is failing, which causes calypso (that requires that package) to fail as well.

Moving the footer to `client` seemed to fix it.

sirbrillig commented on packages/composite-checkout/src/components/checkout-steps.tsx:
```diff
@@ -472,11 +473,13 @@ export function CheckoutStepArea( {
 	children,
 	className,
 	submitButtonHeader,
+	submitButtonFooter,
```
This is a great addition; could you add it to the README? https://github.com/Automattic/wp-calypso/blob/cae8708b131e1790f016a42cf54e50be065fc556/packages/composite-checkout/README.md#L222

keoshi reviewed (approved):
This looks great to me!

The first thing that came to mind when I was testing it is that this badge is visible to all customers purchasing through Calypso _independently of the platform/product_ they use.

I thought this wouldn't apply to dotcom for instance, but it seems like [my assumption was wrong](https://wordpress.com/support/manage-purchases/#refund-policy), so this gets the green light from my side but also pinging dotcom to give them the opportunity to weigh in.

Commit:
Show money back mention only for yearly Jetpack products

Commit:
Update composite checkout README

monsieur-z commented:
> The first thing that came to mind when I was testing it is that this badge is visible to all customers purchasing through Calypso _independently of the platform/product_ they use.

Thanks @keoshi. We've limited the display of that mention to carts that only contain Jetpack products. We'll update the PR to make it 7 days for monthly products.

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx:
```diff
@@ -452,6 +458,14 @@ export default function WPCheckout( {
+					hasYearlyJetpackProductInCart && (
+						<SubmitButtonFooter>
+							<img src={ badge14Src } alt="" />
+							<span>{ translate( '14 day money back guarantee' ) }</span>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( '7 day money back guarantee' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/improve/checkout-money-back-mention)

sirbrillig reviewed (approved):
Looks good to me now! I tested with a Jetpack site (where I see the icon) and on a non-Jetpack site (where I don't see the icon).

I updated the PR description slightly to clarify that this only changes the behavior for Jetpack products.

Worth noting that there is already a 14-day mention in the sidebar of checkout.

Newly added Jetpack product messaging:

<img width="602" alt="Screen Shot 2021-07-05 at 3 45 43 PM" src="https://user-images.githubusercontent.com/2036909/124513288-7610f880-dda8-11eb-85eb-5e25e5342586.png">

Existing messaging for both Jetpack and non-Jetpack products (in sidebar):

<img width="354" alt="Screen Shot 2021-07-05 at 3 48 20 PM" src="https://user-images.githubusercontent.com/2036909/124513309-7f01ca00-dda8-11eb-9b01-7b2236e7c440.png">

sirbrillig commented:
If it's useful, here's the code for the existing money-back guarantee text, which is slightly differently constructed (it looks like it can sometimes be 4 days?):

https://github.com/Automattic/wp-calypso/blob/8de92f4fef7e93b44b40b9b655ee786ea36a514d/client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.js#L144-L161

Commit:
Add 7 day money back guarantee mention

Commit:
Handle case both monthly and yearly products in cart

monsieur-z commented:
I've updated the behaviour and instructions according to the latest specs.

michaeldcain commented:
In general, I really like this fix and could see us implementing it for WPcom Checkout too, but could we tighten up the mobile view? This adds a lot of vertical heigh to the footer, which is fixed-positioned after a user completes the billing information step. There also seems to be some overlap with the inline support link (but maybe that's from an older calypso.live link).

**iPhone X:**
<img width="339" alt="Screen Shot 2021-07-07 at 1 24 49 PM" src="https://user-images.githubusercontent.com/942359/124803582-4b997980-df27-11eb-8992-1524681d704b.png">

With both the header and footer fixed, there's not a lot of room for the credit card form:

**iPhone 6/7/8:**
<img width="339" alt="Screen Shot 2021-07-07 at 1 24 49 PM" src="https://user-images.githubusercontent.com/942359/124803855-9dda9a80-df27-11eb-9b96-7a66cef417e3.gif">

michaeldcain commented:
Just wanted to note that my styling comment is non-blocking. We can follow-up in a separate PR.

sirbrillig commented:
> Just wanted to note that my styling comment is non-blocking. We can follow-up in a separate PR.

although... make sure you can still see the _Total_ line on a mobile device. The video above seems to imply that it might be hidden by the fixed footer, which I think would be a blocker.

sirbrillig commented:
Here's a screenshot of the very bottom of checkout of Chrome emulating an iPhone 4 (so it might not be 100% accurate). Note how the Total is cut off with these changes. Maybe we can adjust whatever padding is underneath the fixed footer so that it still shows?

##### Before this PR

<img width="320" alt="Screen Shot 2021-07-21 at 11 28 26 AM" src="https://user-images.githubusercontent.com/2036909/126516090-379c4ec7-27cd-4e55-8589-01fc686a5886.png">


##### After this PR

<img width="321" alt="Screen Shot 2021-07-21 at 11 27 39 AM" src="https://user-images.githubusercontent.com/2036909/126516115-7f12f16b-f0c8-4ce4-b0f5-5af6624875e2.png">

Commit:
Fix checkout page hidden by footer

Commit:
Fix TS errors

monsieur-z commented:
Nice catch, @sirbrillig, thanks! I updated the code and instructions. When the footer is sticky, the bottom margin of its container matches its height so that there's enough room to display it without hiding the rest of the page. I also rebased the PR along the way.

DavidRothstein commented:
Worth noting (although I realize it's older, and not exactly the same thing): p8Eqe3-G9-p2

sirbrillig commented on packages/composite-checkout/src/components/checkout-steps.tsx:
```diff
@@ -492,11 +495,39 @@ export function CheckoutStepArea( {
+			setSubmitWrapperHeight( submitWrapperRef.current.offsetHeight );
+		}
+		setInitialVw( window.innerWidth );
+	}, [ setSubmitWrapperHeight, submitWrapperRef, setInitialVw ] );
```
I'm pretty sure that `submitWrapperRef` should not be in the dependencies here since it's a Ref and it will never change. Otherwise it implies that the block will re-run when the Ref's value changes.

Based on my understanding of the code, I think that's fine since it looks like this useEffect may only be meant to trigger once, but if you think it needs to trigger when `submitWrapperRef.current` (or `submitWrapperRef.current.offsetHeight`) changes, you can make that a dependency instead (although due to the nature of TypeScript you'll need to assign it to a temporary variable first, like maybe `const submitWrapperHeight = submitWrapper.current?.offsetHeight` and then make `submitWrapperHeight` the dependency).

Commit:
Remove unnecessary deps from React effects

sirbrillig commented:
Thanks for adding that code! A very clever solution. When I load checkout now using a small viewport, the Total is visible!

<img width="321" alt="Screen Shot 2021-07-27 at 10 29 48 AM" src="https://user-images.githubusercontent.com/2036909/127172705-0538aed5-4bf1-4412-9428-43c27a7d1273.png">

The only downside I can see, and this is much less important, is that the solution is not responsive. If I resize my window to be smaller, then the amount of spacing for the footer remains constant. Here's what it looks like if I resize a larger viewport to a smaller one.

<img width="322" alt="Screen Shot 2021-07-27 at 10 29 15 AM" src="https://user-images.githubusercontent.com/2036909/127172894-3e278285-4818-487a-8e0a-8dccde1fdc5f.png">

In practice I'm sure this doesn't happen often at all, so we _could_ probably ignore it, but it does bother me especially since the functionality that manages that spacing is outside of calypso in a soon-to-be published npm package. 🤔 What if we registered a [resize event](https://developer.mozilla.org/en-US/docs/Web/API/Window/resize_event) handler and used it to recalculate the viewport size? If you think that would be a lot more code, I don't want to unnecessarily bloat this PR so it could be a separate change if you like.

Another thing I noticed, that might also be low priority, is that the help button now overlaps the refund text at small widths. 

![overlapping-help-button](https://user-images.githubusercontent.com/2036909/127173806-770d35e8-f299-4f98-b30f-922ce10aa213.png)

I think I remember some effort that had gone into making it not overlap text previously on mobile, leading to its current position, but it's tricky to manage floating elements like that. What do you think?

Commit:
Update checkout page margin bottom on resize

monsieur-z commented:
> it does bother me especially since the functionality that manages that spacing is outside of calypso in a soon-to-be published npm package.

You're right @sirbrillig, especially since the code changes are pretty straightforward.

> Another thing I noticed, that might also be low priority, is that the help button now overlaps the refund text at small widths.

Good catch! I added some padding that should prevent that for the moment. I believe this whole addition will go through another round of design at some point.

sirbrillig reviewed (approved):
Lovely. Thank you for the additions!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6301029

Thank you @monsieur-z for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6301029).
